streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatTransformTest.java

```diff
@@ -0,0 +1,142 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.kstream.TransformerSupplier;
+import org.apache.kafka.streams.kstream.internals.KStreamFlatTransform.KStreamFlatTransformProcessor;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.easymock.EasyMock;
+import org.easymock.EasyMockSupport;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.util.Arrays;
+import java.util.Collections;
+
+import static org.junit.Assert.assertTrue;
+
+public class KStreamFlatTransformTest extends EasyMockSupport {
+
+    private Number inputKey;
+    private Number inputValue;
+
+    private Transformer<Number, Number, Iterable<KeyValue<Integer, Integer>>> transformer;
+    private ProcessorContext context;
+
+    private KStreamFlatTransformProcessor<Number, Number, Integer, Integer> processor;
+
+    @Rule
+    public final ExpectedException exception = ExpectedException.none();
+
+    @Before
+    public void setUp() {
+        inputKey = 1;
+        inputValue = 10;
+        transformer = mock(Transformer.class);
+        context = strictMock(ProcessorContext.class);
+        processor = new KStreamFlatTransformProcessor<Number, Number, Integer, Integer>(transformer);
+    }
+
+    @Test
+    public void shouldInitialiseFlatTransformProcessor() {
+        transformer.init(context);
+        replayAll();
+
+        processor.init(context);
+
+        verifyAll();
+    }
+
+    @Test
+    public void shouldTransformInputRecordToMultipleOutputRecords() {
+        final Iterable<KeyValue<Integer, Integer>> outputRecords = Arrays.asList(
+                KeyValue.pair(2, 20),
+                KeyValue.pair(3, 30),
+                KeyValue.pair(4, 40));
+        processor.init(context);
+        EasyMock.reset(transformer);
+
+        EasyMock.expect(transformer.transform(inputKey, inputValue)).andReturn(outputRecords);
+        for (final KeyValue<Integer, Integer> outputRecord : outputRecords) {
+            context.forward(outputRecord.key, outputRecord.value);
+        }
+        replayAll();
+
+        processor.process(inputKey, inputValue);
+
+        verifyAll();
+    }
+
+    @Test
+    public void shouldAllowEmptyListAsResultOfTransform() {
+        processor.init(context);
+        EasyMock.reset(transformer);
+
+        EasyMock.expect(transformer.transform(inputKey, inputValue))
+            .andReturn(Collections.<KeyValue<Integer, Integer>>emptyList());
+        replayAll();
+
+        processor.process(inputKey, inputValue);
+
+        verifyAll();
+    }
+
+    @Test
+    public void shouldAllowNullAsResultOfTransform() {
```

Comment:
We should add a corresponding test for `flatTransform` in the test above, too. (Positive or negative test, depending if we want to allow `null` or not -- cf. my other comment.)

Reply:
I do not understand to what you refer to? This test tests the `KStreamFlatTransformProcessor`. Both, calls to `transform` and calls to `flatTransform`, run this code. The `transformer` returns a list of output records for both calls. The list can be `null` or it can be empty. The `null`-list is tested in  `shouldAllowNullAsResultOfTransform` and the empty list is tested in `shouldAllowEmptyListAsResultOfTransform`. At this point, there is no distinction between `transform` and `flatTransform`.

Reply:
Ack. My bad.